src/toil/test/spark/sparkTest.py

```diff
@@ -0,0 +1,78 @@
+# Copyright (C) 2016 UC Berkeley AMPLab
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+from toil.common import Toil
+from toil.job import Job
+from toil.spark.spark_cluster import SparkService
+from toil.test import ToilTest, needs_spark
+
+import sys
+
+def _count(job, mem, workers):
+    
+    print >> sys.stderr, "Add spark jobs"
+    # set up cluster
+    masterIP = job.addService(SparkService(False, mem, workers, cores = 1))
+
+    print >> sys.stderr, "Add counting child"
+    job.addChildJobFn(_count_child, masterIP)
+
+def _count_child(job, masterIP):
+
+    print >> sys.stderr, "Starting count"
+
+    # noinspection PyUnresolvedReferences
+    from pyspark import SparkContext
+    
+    # start spark context and connect to cluster
+    sc = SparkContext(master='spark://%s:7077' % masterIP,
+                      appName='count_test')
+
+    print >> sys.stderr, "Creating spark context"
+
+    # create an rdd containing 0-9999 split across 10 partitions
+    rddToWrite = sc.parallelize(xrange(10000), 10)
+
+    # save this rdd to hdfs
+    rddToWrite.saveAsTextFile('hdfs://%s:8020/test.txt' % masterIP)
+
+    # load and count
+    input = sc.textFile('hdfs://%s:8020/test.txt' % masterIP)
+    assert input.count() == 10000
+
+
+class SparkTest(ToilTest):
+
+    def wordCount(badWorker = 0.0):
+
+        # wrap _count as a job
+        countJob = Job.wrapJobFn(_count, 1, 1)
+        options = Job.Runner.getDefaultOptions(self._getTestJobStorePath())
+        options.batchSystem = 'singleMachine'
+        options.badWorker = badWorker
+        options.clean = 'never'
+
+        Job.Runner.startToil(countJob, options)
+    
+    @needs_spark
+    def testSparkLocal(self):
+
+        wordCount()
+
+    @needs_spark
+    def testSparkLocalWithBadWorker(self):
+
+        wordCount(badWorker = 0.5)
```

Comment:
Nice - but if you loop the  testSparkLocalWithBadWorker test a 100 times, I'd expect it to intermittently fail, because you are not using the checkpoint flag for the parent job that defines the spark cluster. Can you confirm that the test does indeed fail from time to time and then check that using the checkpoint function fixes this failure?